aspects/schema.go

```diff
@@ -0,0 +1,328 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2023 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package aspects
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+)
+
+type parser interface {
+	Schema
+
+	// parseConstraints parses constraints for a type defined as a JSON object.
+	// Shouldn't be used with non-object/map type definitions.
+	parseConstraints(json.RawMessage) error
+}
+
+func ParseSchema(raw []byte) (*StorageSchema, error) {
+	var schemaDef map[string]json.RawMessage
+	err := json.Unmarshal(raw, &schemaDef)
+	if err != nil {
+		return nil, fmt.Errorf("cannot parse top level schema: must be a map")
+	}
+
+	schema := &StorageSchema{}
+	// TODO: check "types" here and parse the user-defined types
+
+	if _, ok := schemaDef["schema"]; !ok {
+		return nil, fmt.Errorf(`cannot parse top level schema: must have a "schema" constraint`)
+	}
+
+	schema.topLevel, err = schema.Parse(raw)
+	if err != nil {
+		return nil, err
+	}
+
+	return schema, nil
+}
+
+type StorageSchema struct {
+	// topLevel is the schema for the top level map.
+	topLevel Schema
+}
+
+// Validate validates the provided JSON object
+func (s *StorageSchema) Validate(raw []byte) error {
+	return s.topLevel.Validate(raw)
+}
+
+func (s *StorageSchema) Parse(raw json.RawMessage) (parser, error) {
+	var typ string
+	var schemaDef map[string]json.RawMessage
+	var hasConstraints bool
+	if err := json.Unmarshal(raw, &schemaDef); err != nil {
+		var typeErr *json.UnmarshalTypeError
+		if !errors.As(err, &typeErr) {
+			return nil, err
+		}
+
+		if err := json.Unmarshal(raw, &typ); err != nil {
+			return nil, err
+		}
+	} else {
+		// schema definition is a map, we must process its constraints
+		hasConstraints = true
+
+		rawType, ok := schemaDef["type"]
+		if !ok {
+			typ = "map"
+		} else {
+			if err := json.Unmarshal(rawType, &typ); err != nil {
+				return nil, err
+			}
```

Comment:
I think that's a mix-up between a map type (a schema type that enforces inputs to be maps) and map type declaration (a type declaration that is structured as a JSON object and might include further constraints on that type). At this point, we have type declaration that is structured as a JSON object (not calling it a map just to separate the concepts) but the type might be many things. It could be this:
```
...
{
  "type": "string",
  "pattern": "[fF]oo"
}
```
In which case we'd process this as a string.

Reply:
My comment is in the context of a schema map, in other words a map/object with a schema constraint as per my example. Am I missing something?

Reply:
I'm probably misunderstanding but the example you gave would be processed as a string. Right now, it wouldn't fail because the code is not explicitly disallowing unexpected constraints. But if we decide to do that (which is likely) it will fail because strings don't expect "schema" constraints. I'm not sure what you mean by only maps are allowed to have type map. I thought you meant that only map types can have definitions that are themselves JSON maps, which isn't true. But maybe that's not what you meant. Maybe we can sync face to face

Reply:
Marking as resolved as this is correct (we sync'd on it in a call)